app/src/main/java/org/fossasia/openevent/general/event/EventsViewModel.kt

```diff
@@ -18,32 +23,46 @@ class EventsViewModel(
     private val eventService: EventService,
     private val preference: Preference,
     private val resource: Resource,
-    private val mutableConnectionLiveData: MutableConnectionLiveData
+    private val mutableConnectionLiveData: MutableConnectionLiveData,
+    private val config: PagedList.Config
 ) : ViewModel() {
 
     private val compositeDisposable = CompositeDisposable()
 
     val connection: LiveData<Boolean> = mutableConnectionLiveData
     private val mutableProgress = MutableLiveData<Boolean>()
     val progress: LiveData<Boolean> = mutableProgress
-    private val mutableEvents = MutableLiveData<List<Event>>()
-    val events: LiveData<List<Event>> = mutableEvents
+    private lateinit var mutableEventsPaged: Observable<PagedList<Event>>
+    private val mutableEvents = MutableLiveData<PagedList<Event>>()
+    val events: LiveData<PagedList<Event>> = mutableEvents
     private val mutableError = SingleLiveEvent<String>()
     val error: LiveData<String> = mutableError
     private val mutableShowShimmerEvents = MutableLiveData<Boolean>()
     val showShimmerEvents: LiveData<Boolean> = mutableShowShimmerEvents
     var lastSearch = ""
     private val mutableSavedLocation = MutableLiveData<String>()
     val savedLocation: LiveData<String> = mutableSavedLocation
+    private lateinit var sourceFactory: EventsDataSourceFactory
 
     fun loadLocation() {
         mutableSavedLocation.value = preference.getString(SAVED_LOCATION)
     }
 
     fun loadLocationEvents() {
         if (lastSearch != savedLocation.value) {
-            compositeDisposable += eventService.getEventsByLocation(mutableSavedLocation.value)
-                .withDefaultSchedulers()
+            sourceFactory = EventsDataSourceFactory(
+                compositeDisposable,
+                eventService,
+                mutableSavedLocation.value
+            )
+            mutableEventsPaged = RxPagedListBuilder(sourceFactory, config)
+                .setFetchScheduler(Schedulers.io())
+                .buildObservable()
+                .cache()
+
+            compositeDisposable += mutableEventsPaged
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
                 .doOnSubscribe {
```

Comment:
doFinally Block is not working

Reply:
Cannot get it, why it is not working!

Reply:
@iamareebjamal can you tell why doFinally Block is not working here because of that shimmer layout do not stop everything else just works fine. is it because pagination requires continuous API calls to different pages?

Reply:
@iamareebjamal please a little bit of help here

Reply:
Will try locally today

Reply:
@iamareebjamal any luck with this ?

Reply:
Didn't get time. Hopefully tomorrow

Reply:
@iamareebjamal any update? 😅 i can try to find out someting myself if you're busy

Reply:
Yes, please do. I'm not getting any time these days

Reply:
@iamareebjamal i tried with calling doFinally in the data source and it was working there,i might be right that it RxPagedList don't call doFinally as it subscribes again and again to fetch all the pages

Reply:
i was searching for some solution the only way i found is to send a mutableprogress object to data source and update that on doFinally

Reply:
I've pushed it but I don't think so it is the best possible method to implement it

Reply:
Yeah, doFinally will never be called in case of Observable or Flowable because the stream is never completed